sources/core/input_reader.py

```diff
@@ -8,8 +8,22 @@ class InputReader:
     def read_encrypted_base64_text():
         message = ''
         can_read = False
+        len_header = len(Formatter.HEADER)
         while True:
-            line = input()
+            line = input().strip()
+
+            if Formatter.HEADER in line:
+                if not can_read:
+                    line = line[line.index(Formatter.HEADER) + len_header:]
+                    if Formatter.HEADER in line:
+                        message += line[:line.index(Formatter.HEADER)]
+                        return base64.b64decode(message)
+                    else:
+                        message += line
+                    line = Formatter.HEADER
+                else:
+                    message += line[:line.index(Formatter.HEADER)]
+                    line = Formatter.HEADER
```

Comment:
Hey!

It's 1000 times better, great job! We have a few duplications + logic that are not necessary anymore. Please see my comments below. Feel free to contact me in case of any questions.

```
@staticmethod
    def read_encrypted_base64_text():
        message = ''
        can_read = False
        len_header = len(Formatter.HEADER)
        while True:
            # here we do strip, but later when we extract line, we don't do a strip
            line = input().strip()

            # we do search header in line, but:
            if Formatter.HEADER in line:
                if not can_read:
                    # here we request index of header (the same search)
                    # let's calculate index once and reuse it
                    line = line[line.index(Formatter.HEADER) + len_header:]
                    # here is inner search with inner new index
                    if Formatter.HEADER in line:
                        # (3) - see below
                        message += line[:line.index(Formatter.HEADER)]
                        # return... may be not the best way here ;) but we have break anyway, 
                        # it should be return here ot break there, m?
                        # (6) - see below
                        return base64.b64decode(message)
                    else:
                        # (4) - see below
                        message += line

                    # (1) - see below
                    line = Formatter.HEADER
                else:
                    message += line[:line.index(Formatter.HEADER)]
                    # (2) - see below
                    line = Formatter.HEADER

            # Together with (1) and (2), it seems in general we no need this part of code, it could be
            # simplified.
            if line == Formatter.HEADER:
                if can_read:
                    # didn't check but it seems we never get here, since we return in (6)
                    break
                else:
                    can_read = True
                    continue

            # in general this is NOT of (if Formatter.HEADER in line:)
            # we can optimize it and even remove can_read variable
            if can_read:
                # (5) - see below
                # >> (3, 4, 5) - it seems like a duplication of logic
                message += line

        return base64.b64decode(message)
```